### ISSUE TITLE
LinkedIn Provider does not work for LinkedIn apps created since Aug 2023 (in V4)

### ISSUE THREAD

**Issue by goerlitz** [nextauthjs/next-auth#8956]
### Provider type

LinkedIn

### Environment

```
  System:
    OS: macOS 12.4
    CPU: (10) arm64 Apple M1 Pro
    Memory: 100.25 MB / 16.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 18.17.0 - ~/.nvm/versions/node/v18.17.0/bin/node
    npm: 10.2.0 - ~/.nvm/versions/node/v18.17.0/bin/npm
    pnpm: 7.13.3 - ~/.nvm/versions/node/v18.17.0/bin/pnpm
```

### Reproduction URL

https://github.com/nextauthjs/next-auth-example

### Describe the issue

The LinkedIn provider in `"next-auth": "^4.24.3"` throws errors when it is used with a newly created Linkedin app.

This issue has been raised before, but the problem persists and can only be solved with a workaround. (e.g. see https://github.com/nextauthjs/next-auth/issues/8831)

## Root cause analysis
LinkedIn has 2 different authentication APIs:
1. [Sign in with LinkedIn](https://learn.microsoft.com/en-us/linkedin/consumer/integrations/self-serve/sign-in-with-linkedin) - deprecated as of August 1, 2023, but still in use by older LinkedIn apps
2. [Sign in with LinkedIn using OpenID Connect](https://learn.microsoft.com/en-us/linkedin/consumer/integrations/self-serve/sign-in-with-linkedin-v2) - the new OIDC API that all new LinkedIn dev apps have to use.

LinkedinProvider in NextAuth/Authjs only implements 1 of the 2 APIs.
- the LinkedIn provider in the `v4` branch implements the deprecated LinkedIn authentication API (https://github.com/nextauthjs/next-auth/blob/v4/packages/next-auth/src/providers/linkedin.ts)
- the LinkedIn provider in the `main` branch only implements the new OIDC API (https://github.com/nextauthjs/next-auth/blob/main/packages/core/src/providers/linkedin.ts)

### How to reproduce

Take any example application and
1. created a [new developer app in LinkedIn](https://www.linkedin.com/developers/apps) and use it with the NextAuth 4 LinkedIn Provider
2. use an older LinkedIn developer app (created before Aug 2023) and use it with the NextAuth 5 LinkedIn provider.

### Expected behavior

The LinkedIn provider can be used with both versions of the API to ensure that old and new LinkedIn developer apps can be used.

## Proposed Solution

Either add a configuration option to the LinkedIn provider to choose between old and new API or implement 2 different LinkedIn providers, one for V1 and the other for V2.

**Comment by n1c01a5** [comment]
Any update ?

**Comment by cprussin** [comment]
For anyone else running into this, I have this in my code as a stopgap and it seems to work correctly:

```ts
import type { OAuthConfig } from "next-auth/providers";

const LinkedinProvider = (config: Partial<OAuthConfig<LinkedInProfile>>): OAuthConfig<LinkedInProfile> => ({
  id: "linkedin",
  name: "LinkedIn",
  type: "oauth",
  client: { token_endpoint_auth_method: "client_secret_post" },
  issuer: "https://www.linkedin.com",
  profile: (profile) => ({
    id: profile.sub,
    name: profile.name,
    email: profile.email,
    image: profile.picture,
  }),
  style: { logo: "/linkedin.svg", bg: "#069", text: "#fff" },
  wellKnown: "https://www.linkedin.com/oauth/.well-known/openid-configuration",
  authorization: {
    params: {
      scope: "openid profile email"
    },
  },
  ...config
});
```

I'm going to refrain from putting in a PR though since I just hacked it together and don't really know what I'm doing and someone more knowledgeable than I should assess if it's correct, but it seems to work so maybe it can help unblock others.

**Comment by matheusrocha89** [comment]
Just an update on the @cprussin answer. Thanks @cprussin for this provider, helped me a lot now. I tried the one from the v5 beta and the LinkedIn gave me another error there, but your solution with version 4 worked perfectly.

```typescript
import type { OAuthConfig } from "next-auth/providers/oauth";
import type { LinkedInProfile } from "next-auth/providers/linkedin";

export const LinkedinProvider = (
  config: Partial<OAuthConfig<LinkedInProfile>>
): OAuthConfig<LinkedInProfile> => ({
  id: "linkedin",
  name: "LinkedIn",
  type: "oauth",
  client: { token_endpoint_auth_method: "client_secret_post" },
  issuer: "https://www.linkedin.com",
  profile: (profile: LinkedInProfile) => ({
    id: profile.sub,
    name: profile.name,
    email: profile.email,
    image: profile.picture,
  }),
  wellKnown: "https://www.linkedin.com/oauth/.well-known/openid-configuration",
  authorization: {
    params: {
      scope: "openid profile email",
    },
  },
  style: { logo: "/linkedin.svg", bg: "#069", text: "#fff" },
  ...config,
});
```

**Comment by pierre-H** [comment]
It is linked to https://github.com/nextauthjs/next-auth/issues/8831

**Comment by OsamuBlack** [comment]
Thanks, Linked In Provider for Next Auth no longer works by default due to the removal of r_liteprofile scope.

You can also pass the props directly to the provider.

```typescript
import NextAuth from "next-auth";
import LinkedInProvider, { LinkedInProfile } from "next-auth/providers/linkedin";

const handler = NextAuth({
  providers: [
    LinkedInProvider({
      clientId: process.env.LINKEDIN_CLIENT_ID || "",
      clientSecret: process.env.LINKEDIN_CLIENT_SECRET || "",
      client: { token_endpoint_auth_method: "client_secret_post" },
      issuer: "https://www.linkedin.com",
      profile: (profile: LinkedInProfile) => ({
        id: profile.sub,
        name: profile.name,
        email: profile.email,
        image: profile.picture,
      }),
      wellKnown:
        "https://www.linkedin.com/oauth/.well-known/openid-configuration",
      authorization: {
        params: {
          scope: "openid profile email",
        },
      },
    }),
  ],
});

export { handler as GET, handler as POST };
```

**Comment by goerlitz** [comment]
@OsamuBlack that code was already linked: https://github.com/nextauthjs/next-auth/issues/8831

Moreover, it is not correct that `r_liteprofile` was removed. Instead, there are two different API version provided by LinkedIn. V1 and V2. As already described above.

Please read the ticket description and other comments before posting any code. Thanks!

**Comment by balazsorban44** [comment]
Duplicate of #8831. Will fix when LinkedIn landed the conformity fix.

**Comment by Dragusin-Cristian** [comment]
> For anyone else running into this, I have this in my code as a stopgap and it seems to work correctly:
> 
> ```ts
> import type { OAuthConfig } from "next-auth/providers";
> 
> const LinkedinProvider = (config: Partial<OAuthConfig<LinkedInProfile>>): OAuthConfig<LinkedInProfile> => ({
>   id: "linkedin",
>   name: "LinkedIn",
>   type: "oauth",
>   client: { token_endpoint_auth_method: "client_secret_post" },
>   issuer: "https://www.linkedin.com",
>   profile: (profile) => ({
>     id: profile.sub,
>     name: profile.name,
>     email: profile.email,
>     image: profile.picture,
>   }),
>   wellKnown: "https://www.linkedin.com/oauth/.well-known/openid-configuration",
>   authorization: {
>     params: {
>       scope: "openid profile email"
>     },
>   },
>   ...config
> });
> ```
> 
> I'm going to refrain from putting in a PR though since I just hacked it together and don't really know what I'm doing and someone more knowledgeable than I should assess if it's correct, but it seems to work so maybe it can help unblock others.
> 
> EDIT this is for using next-auth v4.x with new linkedin apps

Very much appreciate it!
